### DISCUSSION TITLE
How to prevent running actions on `push` when `pull_request` is also specified?

### DISCUSSION THREAD

**Discussion by mfn** [community/community#25654] [Actions]
<p>I’m using the classical configuration:</p>
<pre><code class="lang-auto">
name: …

on: [push, pull\_request]

# …

</code></pre>
<p>When a PR is made, this always run both strategies (obvious), but I don’t want that.</p>
<p>What I want is:</p>
<ul>
<li>
<p>as long as there’s not PR, I want to run the build on every push or force-push (for every branch)</p>
</li>
<li>
<p>once there’s a PR (<code>pull_request</code>) I don’t need <code>pull</code> anymore.</p>
</li>
</ul>
<p>Maybe… do I even need <code>pull_request</code> in this scenario?</p>
<p>Basically on every push or force-push, no matter where to (a branch with or without PR) I want the build to run.</p>
<p>This matters for cases where no dedicated PR is needed in branch but I still want to “see” that the tests for this branches are green before merging manually. Or I just want to trigger a test suit run with an idea of a modification.</p>
<p>Maybe I’m overthinking things, not sure. Or I don’t get the proper distinction between them. I looked at <a href="https://help.github.com/en/actions/reference/workflow-syntax-for-github-actions#on" rel="nofollow noopener">https://help.github.com/en/actions/reference/workflow-syntax-for-github-actions#on</a> but can’t derive a clear answer.</p>
<p>It sounds similar to <a href="https://github.community/t/how-to-trigger-an-action-on-push-or-pull-request-but-not-both/16662">https://github.community/t5/GitHub-Actions/How-to-trigger-an-action-on-push-or-pull-request-but-not-both/m-p/35805</a> which doesn’t have a satisfiying answer.</p>
<p>Thanks,</p>
<ul>
<li>Markus</li>
</ul>

**Comment by logankilpatrick** [comment]
<p>Hey! I am trying to understand exactly what you are looking for here. Is it fair to say that your desired workflow is: “Basically on every push or force-push, no matter where to (a branch with or without PR) I want the build to run.” or am I misunderstanding something? </p>

**Comment by Yanjingzhu** [comment]
<blockquote>
<p>What I want is:</p>
</blockquote>
<blockquote>
<ul>
<li>as long as there’s not PR, I want to run the build on every push or force-push (for every branch)</li>
</ul>
</blockquote>
<blockquote>
<ul>
<li>once there’s a PR (<code>pull_request</code>) I don’t need <code>pull</code> anymore.</li>
</ul>
</blockquote>
<p>Let me sort out your requirement, for example there is a pull request from dev to master, when changes push to dev.</p>
<ol>
<li>You don’t want the workflow to be triggered on: push with dev branch. Just on: pull_request is enough.   </li>
</ol>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/5/5139af4b03f87bada7fe19e8be4563c8560089f7.png" data-download-href="/uploads/short-url/bAygn9EVTHY0EGCONvnmsHv9s59.png?dl=1" title="on push pull_request.png" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/24806478/181092372-d37f372e-f5aa-4902-9fdd-44b8eb02089b.png" alt="on push pull_request.png" data-base62-sha1="bAygn9EVTHY0EGCONvnmsHv9s59" width="690" height="95" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/5/5139af4b03f87bada7fe19e8be4563c8560089f7_2_690x95.png, https://aws1.discourse-cdn.com/github/original/2X/5/5139af4b03f87bada7fe19e8be4563c8560089f7.png 1.5x, https://aws1.discourse-cdn.com/github/original/2X/5/5139af4b03f87bada7fe19e8be4563c8560089f7.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/5/5139af4b03f87bada7fe19e8be4563c8560089f7_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">on push pull_request.png</span><span class="informations">966×134 11 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>&gt;&gt; <strong>What I have to say is that your requirement could not be reached.</strong>  </p>
<p>If you want to run a build when push to every branch , you need to specify on:push in the workflow yml file on all branches . This will lead to that on push will work no matter there is a pull request or not. </p>
<p>The workflow on pull request is building a pre merged code from dev to master (refs/pull/id/merge) , the workflow on push is building just code on dev branch. These two workflow runs are running with different codes. You could not prevent the workflow run on push event.    </p>

**Comment by mfn** [comment]
<p>&gt; Hey! I am trying to understand exactly what you are looking for here. Is it fair to say that your desired workflow is: “Basically on every push or force-push, no matter where to (a branch with or without PR) I want the build to run.” or am I misunderstanding something? </p>
<p>Yes, that’s th gist and correct!</p>
<p>&gt; These two workflow runs are running with different codes. You could not prevent the workflow run on push event.</p>
<p>Oh, I think this is the news for me!</p>
<p>So:</p>
<ul>
<li>
<p><code>pull</code> means: check out the “tree on the given commit”, then run action</p>
</li>
<li>
<p>pull_request` means: check out the “tree on given commit and merge with master (or whatever the target is)”, then run action because of what you said:</p>
</li>
</ul>
<p>  &gt; The workflow on pull request is building a pre merged code from dev to master (refs/pull/id/merge)</p>
<p>Did I got this right?</p>
<p>Drive-by question: If so, what happens on merge conflicts with master/target branch?</p>
<p>It totally makes sense but it’s also totally not obvious to me :sweat_smile: I’m not able to find this in the documentation or I’m looking in the wong place.</p>
<p>Places I read:</p>
<ul>
<li><a href="https://help.github.com/en/actions/reference/workflow-syntax-for-github-actions" rel="nofollow noopener">https://help.github.com/en/actions/reference/workflow-syntax-for-github-actions</a></li>
</ul>
<p>  That seems to be the first place to start</p>
<ul>
<li><a href="https://help.github.com/en/actions/reference/events-that-trigger-workflows#pull-request-event-pull_request" rel="nofollow noopener">https://help.github.com/en/actions/reference/events-that-trigger-workflows#pull-request-event-pull_request</a></li>
</ul>
<p>  Very detailed on it’s parameters, but doens’t tell you about the git/tree state?</p>
<p>thanks</p>

**Comment by Yanjingzhu** [comment, marked as answer]
<blockquote>
<ul>
<li>
<code>pull</code> means: check out the “tree on the given commit”, then run action</li>
</ul>
</blockquote>
<blockquote>
<ul>
<li>pull_request` means: check out the “tree on given commit and merge with master (or whatever the target is)”, then run action because of what you said:</li>
</ul>
</blockquote>
<blockquote>
<p>  &gt; The workflow on pull request is building a pre merged code from dev to master (refs/pull/id/merge)</p>
</blockquote>
<blockquote>
<p> </p>
</blockquote>
<blockquote>
<p>Did I got this right?</p>
</blockquote>
<p>Yes, you got it right. The git/tree you mentioned is the GITHUB_REF. You could see the environment variable description <a href="https://help.github.com/en/actions/configuring-and-managing-workflows/using-environment-variables" rel="nofollow noopener">here</a>. </p>
<p><img src="https://user-images.githubusercontent.com/24806478/181092375-03240d7b-31fb-4bd0-9814-fd967d6d1b61.png" alt="Github_ref.png" data-base62-sha1="titxo43f1bjgKw7665UuDX3Knwy" width="429" height="204"></p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/d/d6193a42c7812a76ab2b594867055b7ae2475ea7.png" data-download-href="/uploads/short-url/uy0eaqK93a9vj7pIIXJRNQ6vjb9.png?dl=1" title="PR Ref.png" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/24806478/181092376-71325f8b-f4c0-4b2d-90b0-cfcc74536436.png" alt="PR Ref.png" data-base62-sha1="uy0eaqK93a9vj7pIIXJRNQ6vjb9" width="548" height="500" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/d/d6193a42c7812a76ab2b594867055b7ae2475ea7_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">PR Ref.png</span><span class="informations">761×694 29.6 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div> <div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/4/4f28ff5bbd312c19ab1f14254fc323b7fe69573b.png" data-download-href="/uploads/short-url/bihywGz8wNPXczAVfxKjVxnMuxR.png?dl=1" title="Push ref.png" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/24806478/181092377-3730475f-079c-4d1f-8089-425c6c821bf7.png" alt="Push ref.png" data-base62-sha1="bihywGz8wNPXczAVfxKjVxnMuxR" width="690" height="354" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/4/4f28ff5bbd312c19ab1f14254fc323b7fe69573b_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">Push ref.png</span><span class="informations">757×389 19.3 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>Note:  I think the ‘pull’ you mentioned above is push. </p>
<blockquote>
<p>Drive-by question: If so, what happens on merge conflicts with master/target branch?</p>
</blockquote>
<p>When PR reports merge conflicts, the workflow on: pull_request will not be triggered. </p>

**Comment by mfn** [comment]
<p>Thank you so much, this totally clears things up :bowing_woman:</p>

**Comment by genieai-vikas** [comment]
<p>Have you figure out the answer? It’s happing to me as well. I don’t want multiple builds(for push and PR) when PR is already created</p>
